apps/server-asset-sg/src/features/assets/search/asset-search.writer.ts

```diff
@@ -0,0 +1,202 @@
+import { AssetEditDetail, ElasticPoint, ElasticSearchAsset, GeometryCode, makeUsageCode } from '@asset-sg/shared';
+import { AssetId, ContactId, StudyId, UserId } from '@asset-sg/shared/v2';
+import { Client as ElasticsearchClient } from '@elastic/elasticsearch';
+import { BulkOperationContainer } from '@elastic/elasticsearch/lib/api/types';
+import { PrismaClient } from '@prisma/client';
+import { mapLv95ToElastic } from '@/features/assets/search/asset-search.utils';
+import { StudyRepo } from '@/features/studies/study.repo';
+import { ProcessQueue } from '@/utils/process-queue';
+
+export interface AssetSearchWriterOptions {
+  index: string;
+  shouldRefresh?: boolean;
+  isEager?: boolean;
+}
+
+export class AssetSearchWriter {
+  private readonly eager: Promise<Eager | null>;
+
+  constructor(
+    private readonly elastic: ElasticsearchClient,
+    private readonly prisma: PrismaClient,
+    private readonly studyRepo: StudyRepo,
+    private readonly options: AssetSearchWriterOptions
+  ) {
+    this.eager = options.isEager ? this.fetchEager() : Promise.resolve(null);
+  }
+
+  async write(oneOrMore: AssetEditDetail | AssetEditDetail[]): Promise<void> {
+    const assets = Array.isArray(oneOrMore) ? oneOrMore : [oneOrMore];
+    const operations: Array<BulkOperationContainer | ElasticSearchAsset> = Array(assets.length * 2);
+
+    const processQueue = new ProcessQueue(10);
```

Comment:
I'd move this to either a constant at the beggining of the file (i.e. `const QUEUE_SIZE = 10`) or add it as option. Right now it's a bit buried here.